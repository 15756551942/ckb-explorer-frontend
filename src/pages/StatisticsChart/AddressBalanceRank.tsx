import React, { useEffect, useContext, useCallback } from 'react'
import ReactEchartsCore from 'echarts-for-react/lib/core'
import echarts from 'echarts/lib/echarts'
import 'echarts/lib/chart/bar'
import 'echarts/lib/component/tooltip'
import 'echarts/lib/component/title'
import 'default-passive-events'
import Content from '../../components/Content'
import { getStatisticAddressBalanceRank } from '../../service/app/statisticsChart'
import { StateWithDispatch, PageActions } from '../../contexts/providers/reducer'
import { AppContext } from '../../contexts/providers'
import i18n from '../../utils/i18n'
import Loading from '../../components/Loading'
import { handleAxis } from '../../utils/chart'
import { ChartTitle, ChartPanel, LoadingPanel, ChartCardLoadingPanel } from './styled'
import { isMobile } from '../../utils/screen'
import SmallLoading from '../../components/Loading/SmallLoading'
import { shannonToCkb } from '../../utils/util'
import { localeNumberString } from '../../utils/number'
import { adaptPCEllipsis } from '../../utils/string'
import browserHistory from '../../routes/history'

const colors = ['#3182bd']

const gridThumbnail = {
  left: '4%',
  right: '10%',
  top: '8%',
  bottom: '6%',
  containLabel: true,
}
const grid = {
  left: '3%',
  right: '4%',
  bottom: '3%',
  containLabel: true,
}

const getAddressWithRanking = (statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[], ranking: string) => {
  const addressBalanceRank = statisticAddressBalanceRanks.find(rank => rank.ranking === ranking)
  return addressBalanceRank ? addressBalanceRank.address : ''
}

const getOption = (statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[], isThumbnail = false) => {
  return {
    color: colors,
    tooltip: !isThumbnail && {
      trigger: 'axis',
      formatter: (dataList: any[]) => {
        const colorSpan = (color: string) =>
          `<span style="display:inline-block;margin-right:8px;margin-left:5px;margin-bottom:2px;border-radius:10px;width:6px;height:6px;background-color:${color}"></span>`
        const widthSpan = (value: string) => `<span style="width:60px;display:inline-block;">${value}:</span>`
        let result = `<div>${colorSpan('#333333')}${widthSpan(i18n.t('statistic.address'))} ${adaptPCEllipsis(
          getAddressWithRanking(statisticAddressBalanceRanks, dataList[0].name),
          6,
          60,
        )}</div>`
        result += `<div>${colorSpan(colors[0])}${widthSpan(i18n.t('statistic.balance'))} ${localeNumberString(
          dataList[0].data,
        )} ${i18n.t('common.ckb_unit')}</div>`
        result += `<div>${colorSpan(colors[0])}${widthSpan(i18n.t('statistic.ranking'))} ${dataList[0].name}</div>`
        return result
      },
    },
    grid: isThumbnail ? gridThumbnail : grid,
    xAxis: [
      {
        type: 'category',
        boundaryGap: false,
        data: statisticAddressBalanceRanks.map(data => data.ranking),
      },
    ],
    yAxis: [
      {
        position: 'left',
        name: isMobile() || isThumbnail ? '' : `${i18n.t('statistic.balance_ranking')} ${i18n.t('statistic.log')}`,
        type: 'log',
        logBase: 10,
        scale: true,
        axisLine: {
          lineStyle: {
            color: colors[0],
          },
        },
        axisLabel: {
          formatter: (value: string) => {
            return `${handleAxis(value)}B`
          },
        },
      },
    ],
    series: [
      {
        name: i18n.t('statistic.balance_ranking'),
        type: 'bar',
        yAxisIndex: '0',
        barWidth: '8',
        symbol: isThumbnail ? 'none' : 'circle',
        symbolSize: 3,
        data: statisticAddressBalanceRanks.map(data => shannonToCkb(data.balance)),
      },
    ],
  }
}

export const AddressBalanceRankChart = ({
  statisticAddressBalanceRanks,
  clickEvent,
  isThumbnail = false,
}: {
  statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[]
  clickEvent: any
  isThumbnail?: boolean
}) => {
  if (statisticAddressBalanceRanks.length === 0) {
    return isThumbnail ? (
      <ChartCardLoadingPanel>
        <SmallLoading />
      </ChartCardLoadingPanel>
    ) : null
  }
  return (
    <ReactEchartsCore
      echarts={echarts}
      option={getOption(statisticAddressBalanceRanks, isThumbnail)}
      notMerge
      lazyUpdate
      style={{
        height: isThumbnail ? '230px' : '70vh',
      }}
      onEvents={{
        click: clickEvent,
      }}
    />
  )
}

export default ({ dispatch }: React.PropsWithoutRef<StateWithDispatch>) => {
  const { statisticAddressBalanceRanks } = useContext(AppContext)

  const clickEvent = useCallback(
    (param: any) => {
      if (param && param.name) {
        browserHistory.push(`/address/${getAddressWithRanking(statisticAddressBalanceRanks, param.name)}`)
      }
    },
    [statisticAddressBalanceRanks],
  )

  useEffect(() => {
    dispatch({
      type: PageActions.UpdateStatisticAddressBalanceRank,
      payload: {
        statisticAddressBalanceRanks: [],
      },
    })
    getStatisticAddressBalanceRank(dispatch)
  }, [dispatch])

  return (
    <Content>
      <ChartTitle>{i18n.t('statistic.balance_ranking')}</ChartTitle>
      {statisticAddressBalanceRanks.length > 0 ? (
        <ChartPanel>
          <AddressBalanceRankChart
            statisticAddressBalanceRanks={statisticAddressBalanceRanks}
            clickEvent={clickEvent}
          />
        </ChartPanel>
      ) : (
        <LoadingPanel>
          <Loading show />
        </LoadingPanel>
      )}
    </Content>
  )
}
