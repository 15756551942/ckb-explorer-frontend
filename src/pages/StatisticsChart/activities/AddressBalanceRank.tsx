import React, { useEffect, useCallback } from 'react'
import 'default-passive-events'
import { getStatisticAddressBalanceRank } from '../../../service/app/charts/activities'
import { PageActions, AppDispatch } from '../../../contexts/providers/reducer'
import { useAppState, useDispatch } from '../../../contexts/providers'
import i18n, { currentLanguage } from '../../../utils/i18n'
import { handleAxis } from '../../../utils/chart'
import { isMobile } from '../../../utils/screen'
import { shannonToCkb } from '../../../utils/util'
import { localeNumberString } from '../../../utils/number'
import { adaptPCEllipsis } from '../../../utils/string'
import browserHistory from '../../../routes/history'
import { ChartLoading, ReactChartCore, ChartPage } from '../common/ChartComp'
import { ChartColors } from '../../../utils/const'

const gridThumbnail = {
  left: '4%',
  right: '10%',
  top: '8%',
  bottom: '6%',
  containLabel: true,
}
const grid = {
  left: '5%',
  right: '4%',
  bottom: '5%',
  containLabel: true,
}

const getAddressWithRanking = (statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[], ranking: string) => {
  const addressBalanceRank = statisticAddressBalanceRanks.find(rank => rank.ranking === ranking)
  return addressBalanceRank ? addressBalanceRank.address : ''
}

const getOption = (statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[], isThumbnail = false) => {
  return {
    color: ChartColors,
    tooltip: !isThumbnail && {
      trigger: 'axis',
      formatter: (dataList: any[]) => {
        const colorSpan = (color: string) =>
          `<span style="display:inline-block;margin-right:8px;margin-left:5px;margin-bottom:2px;border-radius:10px;width:6px;height:6px;background-color:${color}"></span>`
        const widthSpan = (value: string) =>
          `<span style="width:${currentLanguage() === 'en' ? '65px' : '35px'};display:inline-block;">${value}:</span>`
        let result = `<div>${colorSpan('#333333')}${widthSpan(i18n.t('statistic.address'))} ${adaptPCEllipsis(
          getAddressWithRanking(statisticAddressBalanceRanks, dataList[0].name),
          6,
          60,
        )}</div>`
        result += `<div>${colorSpan(ChartColors[0])}${widthSpan(i18n.t('statistic.balance'))} ${localeNumberString(
          dataList[0].data,
        )} ${i18n.t('common.ckb_unit')}</div>`
        result += `<div>${colorSpan(ChartColors[0])}${widthSpan(i18n.t('statistic.rank'))} ${dataList[0].name}</div>`
        return result
      },
    },
    grid: isThumbnail ? gridThumbnail : grid,
    xAxis: [
      {
        name: isMobile() || isThumbnail ? '' : i18n.t('statistic.rank'),
        nameLocation: 'middle',
        nameGap: '30',
        type: 'category',
        boundaryGap: false,
        data: statisticAddressBalanceRanks.map(data => data.ranking),
      },
    ],
    yAxis: [
      {
        position: 'left',
        name: isMobile() || isThumbnail ? '' : `${i18n.t('statistic.balance_ranking')} ${i18n.t('statistic.log')}`,
        type: 'log',
        logBase: 10,
        scale: true,
        axisLine: {
          lineStyle: {
            color: ChartColors[0],
          },
        },
        axisLabel: {
          formatter: (value: string) => {
            return `${handleAxis(value)}B`
          },
        },
      },
    ],
    series: [
      {
        name: i18n.t('statistic.balance_ranking'),
        type: 'bar',
        yAxisIndex: '0',
        barWidth: '8',
        symbol: isThumbnail ? 'none' : 'circle',
        symbolSize: 3,
        data: statisticAddressBalanceRanks.map(data => shannonToCkb(data.balance)),
      },
    ],
  }
}

export const AddressBalanceRankChart = ({
  statisticAddressBalanceRanks,
  clickEvent,
  isThumbnail = false,
}: {
  statisticAddressBalanceRanks: State.StatisticAddressBalanceRank[]
  clickEvent: any
  isThumbnail?: boolean
}) => {
  if (!statisticAddressBalanceRanks || statisticAddressBalanceRanks.length === 0) {
    return <ChartLoading show={statisticAddressBalanceRanks === undefined} isThumbnail={isThumbnail} />
  }
  return (
    <ReactChartCore
      option={getOption(statisticAddressBalanceRanks, isThumbnail)}
      isThumbnail={isThumbnail}
      clickEvent={clickEvent}
    />
  )
}

export const initStatisticAddressBalanceRanks = (dispatch: AppDispatch) => {
  dispatch({
    type: PageActions.UpdateStatisticAddressBalanceRank,
    payload: {
      statisticAddressBalanceRanks: undefined,
    },
  })
}

export default () => {
  const dispatch = useDispatch()
  const { statisticAddressBalanceRanks } = useAppState()

  const clickEvent = useCallback(
    (param: any) => {
      if (param && param.name) {
        browserHistory.push(`/address/${getAddressWithRanking(statisticAddressBalanceRanks, param.name)}`)
      }
    },
    [statisticAddressBalanceRanks],
  )

  useEffect(() => {
    initStatisticAddressBalanceRanks(dispatch)
    getStatisticAddressBalanceRank(dispatch)
  }, [dispatch])

  return (
    <ChartPage title={i18n.t('statistic.balance_ranking')}>
      <AddressBalanceRankChart statisticAddressBalanceRanks={statisticAddressBalanceRanks} clickEvent={clickEvent} />
    </ChartPage>
  )
}
