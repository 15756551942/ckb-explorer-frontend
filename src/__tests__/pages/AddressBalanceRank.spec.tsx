import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import AddressBalanceRank from '../../pages/StatisticsChart/activities/AddressBalanceRank'

describe('Active Page', () => {
  it('is defined', () => {
    const app = shallow(<AddressBalanceRank />)

    expect(app).toBeDefined()
  })

  it('shallow renders with empty data', async () => {
    const wrapper = renderer.create(<AddressBalanceRank />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })
})
