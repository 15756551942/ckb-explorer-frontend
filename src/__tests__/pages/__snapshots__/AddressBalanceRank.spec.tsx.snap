// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active Page shallow renders with empty data 1`] = `
<div
  className="sc-fznxsB cFeRni"
>
  <div
    className="sc-AxirZ fPaPCL"
  >
    Balance Ranking
  </div>
  <div
    className="sc-AxjAm eHANUb"
  >
    <div
      className="sc-AxiKw dInfRO"
    >
      <div
        className="sc-AxhCb bEWJpg"
      >
        <img
          alt="no data"
          src="chart_no_data.png"
        />
        <span>
          No data...
        </span>
      </div>
    </div>
  </div>
</div>
`;
